File: android/src/main/java/com/reactnativecommunity/webview/RNCGigya.kt
package com.reactnativecommunity.webview

import android.app.Application
import android.webkit.WebResourceRequest
import android.webkit.WebView
import android.webkit.WebViewClient
import com.gigya.android.sdk.Gigya
import com.gigya.android.sdk.GigyaPluginCallback
import com.gigya.android.sdk.account.models.GigyaAccount
import com.gigya.android.sdk.session.SessionInfo
import com.gigya.android.sdk.ui.plugin.IGigyaWebBridge

class RNCGigya(context: Application, apiKey: String, apiDomain: String) {
  private var gigya: Gigya<GigyaAccount>

  init {
    Gigya.setApplication(context)
    gigya = Gigya.getInstance(GigyaAccount::class.java)
    gigya.init(apiKey, apiDomain)
  }

  fun initialize(sessionToken: String, sessionSecret: String, webview: WebView) {
    attachBridge(webview)
    logUser(sessionToken = sessionToken, sessionSecret = sessionSecret)
  }

  private fun logUser(sessionToken: String, sessionSecret: String) {
    val session = SessionInfo(sessionSecret, sessionToken)
    gigya.setSession(session)
  }


  private fun attachBridge(webview: WebView) {
    var webBridge: IGigyaWebBridge<GigyaAccount>? = null

    /*
    Make sure you enable javascript for your WebView instance.
    */
    val webSettings = webview.settings
    webSettings.javaScriptEnabled = true

    webBridge = gigya.createWebBridge()
    webBridge?.attachTo(webview, object: GigyaPluginCallback<GigyaAccount>() {}, null)

    /*
    Make sure to attach the GigyaWebBridge to your WebViewClient instance.
    */
    webview.webViewClient = (object: WebViewClient() {
      override fun shouldOverrideUrlLoading(view: WebView?, request: WebResourceRequest?): Boolean {
        val uri = request?.url
        val uriString = uri.toString()
        return webBridge?.invoke(uriString) ?: false
      }
    })
  }
}

File: apple/RNCGigya.swift
//
//  RNCGigya.swift
//  RNCWebView
//
//  Created by Martin Frouin on 19/10/2023.
//  Copyright © 2023 Facebook. All rights reserved.
//

import Foundation
import UIKit
import WebKit
import Gigya

@objc
class RNCGigya: NSObject {
  
  @objc
  static let shared = RNCGigya()
  
  func initialize(controller: UIViewController, webview: WKWebView, sessionToken: String, sessionSecret: String, apiKey: String, apiDomain: String) {
    gigya.initFor(apiKey: apiKey, apiDomain: apiDomain)

    self.attachBridge(controller: controller, webview: webview)
    self.logUser(sessionToken: sessionToken, sessionSecret: sessionSecret)
  }
  
  func attachBridge(controller: UIViewController, webview: WKWebView) {
      let webBridge = gigya.createWebBridge()
      
      webBridge.attachTo(webView: webview, viewController: controller) { _ in }
  
  }

  func logUser(sessionToken: String, sessionSecret: String) {
      let session = GigyaSession(sessionToken: sessionToken, secret: sessionSecret)
      
      if (session != nil) {
          gigya.setSession(session!)
      }
  }
  
}
